src/memory.rs

```diff
@@ -42,22 +45,37 @@ impl Pointer {
 // Top-level interpreter memory
 ////////////////////////////////////////////////////////////////////////////////
 
-pub struct Memory {
+pub struct Memory<'tcx> {
     alloc_map: HashMap<AllocId, Allocation>,
+    functions: HashMap<AllocId, (DefId, &'tcx Substs<'tcx>)>,
```

Comment:
Add a comment that `alloc_map`'s and `functions`'s key sets are mutually exclusive (i.e. no `AllocId` is present in both maps), since that's a fairly important rule for catching user code that transmutes between fn ptrs and other ptrs.